chainer/functions/array/permutate.py

```diff
@@ -0,0 +1,89 @@
+import numpy
+import six
+
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+def _check_indices(indices):
+    if len(indices) == 0:
+        return
+    for i in indices:
+        if 0 <= i < len(indices):
+            continue
+        raise ValueError('Out of bounds index: {}'.format(i))
+    sort = numpy.sort(indices)
+    for s, t in six.moves.zip(sort, sort[1:]):
+        if s == t:
+            raise ValueError('indices contains duplicate value: {}'.format(s))
+
+
+def _reverse_indices(indices):
+    r = numpy.empty(len(indices), 'i')
+    for i, ind in enumerate(indices):
+        r[ind] = i
+    return r
+
+
+class Permutate(function.Function):
+
+    """Permutate function."""
+
+    def __init__(self, indices, axis=0, rev=False):
+        _check_indices(indices)
```

Comment:
Could checking the validity of indices be the performance bottleneck? If so, I recommend to invoke the check only when debug mode.